### ISSUE TITLE
Copy-paste code while teleporting a player

### ISSUE THREAD

**Issue by NiLSPACE** [cuberite/Core#173]
Every time we want to teleport a player we use cWorld:ChunkStay, but that causes allot of copy-paste code. Instead I think we should override the cPlayer.TeleportToCoords function. Something like this:

``` lua
do
    local teleportFunction = cPlayer.TeleportToCoords
    cPlayer.TeleportToCoords = function(a_Player, a_PosX, a_PosY, a_PosZ)
        local world = a_Player:GetWorld()
        local playeruuid = a_Player:GetUUID()
        world:ChunkStay({{math.floor(a_PosX / 16), math.floor(a_PosZ / 16)}}, 
            function()
                world:DoWithPlayerByUUID(playeruuid, 
                    function(a_Player)
                        teleportFunction(a_Player, a_PosX, a_PosY, a_PosZ)
                    end
                )
            end
        )
    end
end
```

**Comment by madmaxoft** [comment]
Bam, and you made the same error that everyone else makes and we have to look out for it. You must not use `a_Player` in the ChunkStay's callback, because by that time the player may have disconnected.

This is exactly why we need the automated plugin testing :)

**Comment by NiLSPACE** [comment]
Good point. On the other hand, if we'd fix it there we'd fix it everywhere (in Core) :)

**Comment by NiLSPACE** [comment]
Would using the cPlayer object cause a guaranteed crash, or only when we'd access it while the player had left already?

**Comment by madmaxoft** [comment]
The code will work okay as long as the `cPlayer` object is valid - which is what makes this error dangerous, because it's so hidden and normally untestable.

**Comment by LogicParrot** [comment]
I don't think we need `ChunkStay` at all any more. I prevented players from falling into unloaded chunks back in April, so `ChunkStay` is redundant. Just teleport the player and you should be fine, I think.

**Comment by NiLSPACE** [comment]
If that was implemented I guess ChunkStay isn't really needed anymore for teleporting players

**Comment by madmaxoft** [comment]
I still think it's better to first load the chunk and then move the player.

**Comment by NiLSPACE** [comment]
But then the player might think the server is really lagging behind since it can't even teleport the player directly after the command (like in vanilla)

**Comment by LogicParrot** [comment]
Plus, instant teleport is important for game logic. By the time your "/sos-teleport" command teleports you on a high-load day, the dragon would have killed you.

**Comment by madmaxoft** [comment]
Do you really think waiting for 1 chunk to load would produce such a terrible lag?

**Comment by LogicParrot** [comment]
If I remember my experience on @jammet's "Lonely Island" correctly, Chunks can take several seconds to load sometimes. @jammet?

**Comment by LogicParrot** [comment]
Worth noting it's a Raspberry Pi server.

**Comment by LogicParrot** [comment]
If this sounds insane, https://github.com/cuberite/cuberite/issues/2905 may be related. Specifically, the part about wrong chunk load priorities.

**Comment by LogicParrot** [comment]
Also note that "loading" in this case means "loading and sending to client", so if someone is torrenting while playing, or if their connection speed drops for some reason, or if it's naturally slow, this will easily happen on a performant server.

**Comment by jammet** [comment]
Yes, chunk loading can be fairly slow sometimes. It's not uncommon to stand in one place for half a minute and just wait. People have gotten used to this, and it still plays pleasantly. No one is travelling at very high speeds, and chunk loading is largely unnoticed. When teleporting, you notice how it's not reliable though, and that chunks loading seems to have absolutely no priority, as you'll see a world with gaping holes in it everywhere, some right in front of you, while the chunks in the distance have already loaded. I love having the server on the pi though. It's amazingly fun.

**Comment by tigerw** [comment]
> I don't think we need ChunkStay at all any more.

It is needed for ensuring chunks remain loaded and ensuring chunks are present before data about them are requested.

**Comment by LogicParrot** [comment]
Sorry Tiger, let me rephrase:
We don't need to use Chunkstay when teleporting players. It is still needed in general, for other stuff.

**Comment by tigerw** [comment]
OK.

What does vanilla do?

**Comment by williamhatcher** [comment]
Maybe remove the `easy` tag?

**Comment by LogicParrot** [comment]
Done

**Comment by madmaxoft** [comment]
So if ChunkStay is not needed anymore, is this issue even relevant? Can it be closed?

**Comment by LogicParrot** [comment]
Probably
